File: examples/generators/native-go/db_config.go
// Package usergens provides example native generators for Go custom builds.
//
// Usage: place this file at .shatter/generators/db_config.go and reference it
// in config.yaml:
//
//	defaults:
//	  generators:
//	    DbConfig: .shatter/generators/db_config.go
//
// Then build: shatter build-frontend go
package usergens

import (
	"encoding/json"
	"fmt"
	"math/rand"

	"github.com/shatter-dev/shatter/shatter-go/generators"
)

// DbConfig generates database configuration objects. On fresh generation
// (recipe == nil), it creates a randomized test database name. On replay,
// it reconstructs from the stored recipe.
func DbConfig(recipe json.RawMessage) generators.NativeGeneratorResult {
	type Config struct {
		Host string `json:"host"`
		Port int    `json:"port"`
		DB   string `json:"db"`
	}

	var cfg Config
	if recipe != nil {
		_ = json.Unmarshal(recipe, &cfg)
	} else {
		cfg = Config{
			Host: "localhost",
			Port: 5432,
			DB:   fmt.Sprintf("test_%04d", rand.Intn(10000)),
		}
	}

	recipeBytes, _ := json.Marshal(cfg)
	return generators.NativeGeneratorResult{
		ID:     "local-postgres",
		Value:  cfg, // live object, stays in-process
		Recipe: recipeBytes,
	}
}

File: examples/generators/native-rust/db_config.rs
// Example native generator for Rust custom builds.
//
// Usage: place at .shatter/generators/db_config.rs and reference in config.yaml:
//
//   defaults:
//     generators:
//       DbConfig: .shatter/generators/db_config.rs
//
// Then build: shatter build-frontend rust

use shatter_rust::generators::GeneratorResult;

/// DbConfig generates database configuration objects. On fresh generation
/// (recipe is None), it creates a randomized test database name. On replay,
/// it reconstructs from the stored recipe.
pub fn DbConfig(recipe: Option<serde_json::Value>) -> GeneratorResult {
    let cfg = match recipe {
        Some(r) => r,
        None => {
            let db_name = format!("test_{:04}", rand::random::<u16>() % 10000);
            serde_json::json!({
                "host": "localhost",
                "port": 5432,
                "db": db_name
            })
        }
    };

    GeneratorResult {
        id: "local-postgres".into(),
        value: Box::new(cfg.clone()),
        recipe: cfg,
    }
}
